paych/lotus/module.go

```diff
@@ -0,0 +1,178 @@
+	"github.com/ipfs/go-cid"
+	"github.com/textileio/powergate/ffs"
+)
+
+// Module provides access to the paych api
+type Module struct {
+	api *apistruct.FullNodeStruct
+}
+
+var _ ffs.PaychManager = (*Module)(nil)
+
+// New creates a new paych module
+func New(api *apistruct.FullNodeStruct) *Module {
+	return &Module{
+		api: api,
+	}
+}
+
+// List lists all payment channels involving the specified addresses
+func (m *Module) List(ctx context.Context, addrs ...string) ([]ffs.PaychInfo, error) {
+	filter := make(map[string]struct{}, len(addrs))
+	for _, addr := range addrs {
+		filter[addr] = struct{}{}
+	}
+
+	allAddrs, err := m.api.PaychList(ctx)
+	if err != nil {
+		return nil, err
+	}
+
+	chans := make([]<-chan statusResult, len(allAddrs))
+	for i, addr := range allAddrs {
+		chans[i] = m.paychStatus(ctx, addr)
+	}
+
+	resultsCh := make(chan statusResult, len(chans))
+	for _, c := range chans {
+		go func(ch <-chan statusResult) {
+			res := <-ch
+			resultsCh <- res
+		}(c)
+	}
+
+	results := make([]statusResult, len(chans))
+	for i := 0; i < len(chans); i++ {
+		results[i] = <-resultsCh
+	}
+
+	var final []ffs.PaychInfo
+	for _, result := range results {
+		if result.err != nil {
+			// ToDo: do we want to fail if there was an error checking
+			// even one status that may not even be in our filter?
+			continue
+		}
+		_, addrInFilter := filter[result.addr.String()]
+		_, ctlAddrInFilter := filter[result.status.ControlAddr.String()]
+		if addrInFilter || ctlAddrInFilter {
+			var dir ffs.PaychDir
+			switch result.status.Direction {
+			case api.PCHUndef:
+				dir = ffs.PaychDirUnspecified
+			case api.PCHInbound:
+				dir = ffs.PaychDirInbound
+			case api.PCHOutbound:
+				dir = ffs.PaychDirOutbound
+			default:
+				return nil, fmt.Errorf("unknown pay channel direction %v", result.status.Direction)
+			}
+
+			info := ffs.PaychInfo{
+				CtlAddr:   result.status.ControlAddr.String(),
+				Addr:      result.addr.String(),
+				Direction: dir,
+			}
+
+			final = append(final, info)
+		}
+	}
+
+	return final, nil
+}
+
+// Create creates a new payment channel
+func (m *Module) Create(ctx context.Context, from string, to string, amount uint64) (ffs.PaychInfo, cid.Cid, error) {
+	return ffs.PaychInfo{}, cid.Undef, fmt.Errorf("unimplemeted for now, blocked by lotus issue #1611")
+	// fAddr, err := address.NewFromString(from)
+	// if err != nil {
+	// 	return ffs.PaychInfo{}, cid.Undef, err
+	// }
+	// tAddr, err := address.NewFromString(from)
+	// if err != nil {
+	// 	return ffs.PaychInfo{}, cid.Undef, err
+	// }
+	// a := types.NewInt(amount)
+	// info, err := m.api.PaychGet(ctx, fAddr, tAddr, a)
+	// if err != nil {
+	// 	return ffs.PaychInfo{}, cid.Undef, err
+	// }
+	// // ToDo: verify these addresses and direction make sense
+	// res := ffs.PaychInfo{
+	// 	CtlAddr:   from,
+	// 	Addr:      info.Channel.String(),
+	// 	Direction: ffs.PaychDirOutbound,
+	// }
+	// return res, info.ChannelMessage, nil
```

Comment:
Interestingly, it seems that the `info.Channel` address is the pay channel address. I _think_ this is NOT the address you set up the pay channel "to". Strangely, it seems that after you initially create the pay channel, the "to" address is obscured away in the pay channel from then on. Will lead to some weird UX in a cli or UI. Takeaway is that pay channels might be more of an internal tool than the UI work wants them to be.

Reply:
Yeah, I guess will be discovered with time.